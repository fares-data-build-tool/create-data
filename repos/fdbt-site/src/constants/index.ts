export const OPERATOR_COOKIE: string = 'fdbt-operator-cookie';

export const SERVICE_COOKIE: string = 'fdbt-service-cookie';

export const FARETYPE_COOKIE: string = 'fdbt-faretype-cookie';

export const FEEDBACK_LINK: string = 'mailto:fdbt@transportforthenorth.com';

export const GOVUK_LINK: string = 'https://www.nationalarchives.gov.uk/doc/open-government-licence/version/3/';

export const CSV_INPUT_METHOD_COOKIE: string = 'fdbt-csv-input-method-cookie';

export const MANUAL_INPUT_METHOD_COOKIE: string = 'fdbt-manual-input-method-cookie';

export const INT_MAP_INPUT_METHOD_COOKIE: string = 'fdbt-interactive-map-input-method-cookie';
