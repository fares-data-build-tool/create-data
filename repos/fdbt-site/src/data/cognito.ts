import { CognitoIdentityServiceProvider } from 'aws-sdk';
import awsParamStore from 'aws-param-store';
import crypto from 'crypto';

const clientId = process.env.FDBT_USER_POOL_CLIENT_ID as string;
const userPoolId = process.env.FDBT_USER_POOL_ID as string;
let clientSecret: string | undefined;

const getCognitoClient = (): CognitoIdentityServiceProvider =>
    new CognitoIdentityServiceProvider({ region: 'eu-west-2' });

const cognito = getCognitoClient();

const calculateSecretHash = (username: string): string => {
    if (!clientSecret) {
        clientSecret = awsParamStore.getParameterSync('fdbt-cognito-client-secret', { region: 'eu-west-2' }).Value;
    }

    return crypto
        .createHmac('SHA256', clientSecret as string)
        .update(username + clientId)
        .digest('base64');
};

export const initiateAuth = async (
    username: string,
    password: string,
): Promise<CognitoIdentityServiceProvider.AdminInitiateAuthResponse> => {
    console.info('initiating auth...');

    const params: CognitoIdentityServiceProvider.AdminInitiateAuthRequest = {
        AuthFlow: 'ADMIN_USER_PASSWORD_AUTH',
        ClientId: clientId,
        UserPoolId: userPoolId,
        AuthParameters: {
            USERNAME: username,
            PASSWORD: password,
            SECRET_HASH: calculateSecretHash(username),
        },
    };

    try {
        const response = await cognito.adminInitiateAuth(params).promise();

        return response;
    } catch (error) {
        throw new Error(`Failed to authenticate user: ${error.stack}`);
    }
};

export const initiateRefreshAuth = async (
    username: string,
    refreshToken: string,
): Promise<CognitoIdentityServiceProvider.AdminInitiateAuthResponse> => {
    console.info('initiating refresh auth...');

    const params: CognitoIdentityServiceProvider.AdminInitiateAuthRequest = {
        AuthFlow: 'REFRESH_TOKEN_AUTH',
        ClientId: clientId,
        UserPoolId: userPoolId,
        AuthParameters: {
            REFRESH_TOKEN: refreshToken,
            SECRET_HASH: calculateSecretHash(username),
        },
    };

    try {
        const response = await cognito.adminInitiateAuth(params).promise();

        return response;
    } catch (error) {
        throw new Error(`Failed to refresh user session: ${error.stack}`);
    }
};

export const respondToNewPasswordChallenge = async (
    username: string,
    password: string,
    session: string,
): Promise<void> => {
    console.info('new password challenge initiated...');

    const params: CognitoIdentityServiceProvider.AdminRespondToAuthChallengeRequest = {
        ChallengeName: 'NEW_PASSWORD_REQUIRED',
        ClientId: clientId,
        UserPoolId: userPoolId,
        ChallengeResponses: {
            USERNAME: username,
            NEW_PASSWORD: password,
            SECRET_HASH: calculateSecretHash(username),
        },
        Session: session,
    };

    try {
        await cognito.adminRespondToAuthChallenge(params).promise();
    } catch (error) {
        throw new Error(`Failed to respond to password challenge: ${error.stack}`);
    }
};

export const globalSignOut = async (username: string): Promise<void> => {
    console.info('performing global sign out...');

    const params: CognitoIdentityServiceProvider.AdminUserGlobalSignOutRequest = {
        Username: username,
        UserPoolId: userPoolId,
    };

    try {
        await cognito.adminUserGlobalSignOut(params).promise();
    } catch (error) {
        throw new Error(`Failed to perform global sign out: ${error.stack}`);
    }
};

export const updateUserAttributes = async (
    username: string,
    attributes: { Name: string; Value: string }[],
): Promise<void> => {
    console.info('updating user attributes...');

    const params: CognitoIdentityServiceProvider.AdminUpdateUserAttributesRequest = {
        UserAttributes: attributes,
        UserPoolId: userPoolId,
        Username: username,
    };

    try {
        await cognito.adminUpdateUserAttributes(params).promise();
    } catch (error) {
        throw new Error(`Failed to update user attributes: ${error.stack}`);
    }
};

export const forgotPassword = async (username: string): Promise<void> => {
    console.info('start forgot password...');

    const params: CognitoIdentityServiceProvider.ForgotPasswordRequest = {
        ClientId: clientId,
        Username: username,
        SecretHash: calculateSecretHash(username),
    };

    try {
        await cognito.forgotPassword(params).promise();
    } catch (error) {
        throw new Error(`Failed to perform forgot password request: ${error.stack}`);
    }
};

export const confirmForgotPassword = async (
    username: string,
    confirmationCode: string,
    password: string,
): Promise<void> => {
    console.info('confirm forgot password...');

    const params: CognitoIdentityServiceProvider.ConfirmForgotPasswordRequest = {
        ClientId: clientId,
        Username: username,
        ConfirmationCode: confirmationCode,
        Password: password,
        SecretHash: calculateSecretHash(username),
    };

    try {
        await cognito.confirmForgotPassword(params).promise();
    } catch (error) {
        if (error?.code === 'ExpiredCodeException') {
            throw new Error('ExpiredCodeException');
        }
        throw new Error(`Failed to confirm forgotten password: ${error.stack}`);
    }
};

export const updateUserPassword = async (newPassword: string, username: string): Promise<void> => {
    console.info('updating user password...');

    const params: CognitoIdentityServiceProvider.AdminSetUserPasswordRequest = {
        Password: newPassword,
        Permanent: true,
        Username: username,
        UserPoolId: userPoolId,
    };

    try {
        await cognito.adminSetUserPassword(params).promise();
    } catch (error) {
        throw new Error(`Failed to update user password: ${error.stack}`);
    }
};
