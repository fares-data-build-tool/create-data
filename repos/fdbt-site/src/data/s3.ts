import AWS from 'aws-sdk';

export interface FareStage {
    stageName: string;
    prices: {
        price: string;
        fareZones: string[];
    }[];
}

export interface UserFareStages {
    fareStages: FareStage[];
}

export interface UserFareZone {
    FareZoneName: string;
    NaptanCodes: string;
    AtcoCodes: string;
}

const getS3Client = (): AWS.S3 => {
    let options = {};

    if (process.env.NODE_ENV === 'development') {
        options = {
            s3ForcePathStyle: true,
            accessKeyId: 'S3RVER',
            secretAccessKey: 'S3RVER',
            endpoint: new AWS.Endpoint('http://localhost:4572'),
        };
    }

    return new AWS.S3(options);
};

const s3 = getS3Client();

export const getUserFareStages = async (uuid: string): Promise<UserFareStages> => {
    if (!process.env.USER_DATA_BUCKET_NAME) {
        throw new Error('Environment variable for validated bucket not set');
    }

    const params = {
        Bucket: process.env.USER_DATA_BUCKET_NAME,
        Key: `${uuid}.json`,
    };
    const response = await s3.getObject(params).promise();
    const dataAsString = response.Body?.toString('utf-8') ?? '';

    return JSON.parse(dataAsString);
};

export const putStringInS3 = async (
    bucketName: string,
    key: string,
    text: string,
    contentType: string,
): Promise<void> => {
    const request: AWS.S3.Types.PutObjectRequest = {
        Bucket: bucketName,
        Key: key,
        Body: Buffer.from(text, 'binary'),
        ContentType: contentType,
    };

    await s3.putObject(request).promise();
};

export const putDataInS3 = async (
    data: UserFareZone[] | UserFareStages | string,
    key: string,
    processed: boolean,
): Promise<void> => {
    let contentType = '';
    let bucketName = '';

    if (!process.env.USER_DATA_BUCKET_NAME || !process.env.RAW_USER_DATA_BUCKET_NAME) {
        throw new Error('Bucket name environment variables not set.');
    }

    if (processed) {
        bucketName = process.env.USER_DATA_BUCKET_NAME;
        contentType = 'application/json; charset=utf-8';
    } else {
        bucketName = process.env.RAW_USER_DATA_BUCKET_NAME;
        contentType = 'text/csv; charset=utf-8';
    }

    await putStringInS3(bucketName, key, JSON.stringify(data), contentType);
};
