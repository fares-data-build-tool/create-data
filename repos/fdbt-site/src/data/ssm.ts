import { GetParameterCommand, SSMClient } from '@aws-sdk/client-ssm';

const ssm = new SSMClient({ region: 'eu-west-2' });

export const getSsmValue = async (name: string, withDecryption?: boolean): Promise<string> => {
    const input = {
        Name: name,
        WithDecryption: withDecryption ? withDecryption : true,
    };
    const value = (await ssm.send(new GetParameterCommand(input))).Parameter?.Value;

    if (!value) {
        throw new Error(`Parameter ${name} was not found`);
    }

    return value;
};
