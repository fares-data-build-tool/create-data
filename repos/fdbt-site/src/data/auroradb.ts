import dateFormat from 'dateformat';
import { createPool, Pool } from 'mysql2/promise';
import awsParamStore from 'aws-param-store';
import logger from '../utils/logger';
import { SalesOfferPackage } from '../pages/api/describeSalesOfferPackage';
import { INTERNAL_NOC } from '../constants';

export interface ServiceType {
    lineName: string;
    startDate: string;
    description: string;
    serviceCode: string;
}

export interface OperatorNameType {
    operatorPublicName: string;
    nocCode?: string;
}

export interface Operator {
    operatorPublicName: string;
    nocCode: string;
}

export interface JourneyPattern {
    startPoint: {
        Id: string;
        Display: string;
    };
    endPoint: {
        Id: string;
        Display: string;
    };
    stopList: string[];
}

export interface QueryData {
    operatorShortName: string;
    serviceDescription: string;
    lineName: string;
    fromAtcoCode: string;
    toAtcoCode: string;
    fromCommonName: string;
    toCommonName: string;
    journeyPatternId: string;
    order: string;
}
export interface RawJourneyPattern {
    orderedStopPoints: {
        stopPointRef: string;
        commonName: string;
    }[];
}

interface NaptanInfo {
    commonName: string;
    naptanCode: string;
    atcoCode: string;
    nptgLocalityCode: string;
    localityName: string;
    parentLocalityName: string;
    indicator: string;
    street: string;
}

interface NaptanAtcoCodes {
    naptanCode: string;
    atcoCode: string;
}

export interface Stop {
    stopName: string;
    naptanCode: string;
    atcoCode: string;
    localityCode: string;
    localityName: string;
    parentLocalityName: string;
    indicator?: string;
    street?: string;
    qualifierName?: string;
}

export interface StopIdentifiers {
    naptanCode: string | null;
    atcoCode: string;
}

export interface Service {
    serviceDescription: string;
    operatorShortName: string;
    journeyPatterns: JourneyPattern[];
}

export interface RawService {
    serviceDescription: string;
    operatorShortName: string;
    journeyPatterns: RawJourneyPattern[];
}

interface RawSalesOfferPackage {
    name: string;
    description: string;
    purchaseLocations: string;
    paymentMethods: string;
    ticketFormats: string;
}

export const getAuroraDBClient = (): Pool => {
    let client: Pool;

    if (process.env.NODE_ENV === 'development' || process.env.NODE_ENV === 'test') {
        client = createPool({
            host: 'localhost',
            user: 'fdbt_site',
            password: 'password',
            database: 'fdbt',
            waitForConnections: true,
            connectionLimit: 10,
            queueLimit: 0,
        });
    } else {
        client = createPool({
            host: process.env.RDS_HOST,
            user: awsParamStore.getParameterSync('fdbt-rds-site-username', { region: 'eu-west-2' }).Value,
            password: awsParamStore.getParameterSync('fdbt-rds-site-password', { region: 'eu-west-2' }).Value,
            database: 'fdbt',
            waitForConnections: true,
            connectionLimit: 10,
            queueLimit: 0,
        });
    }

    return client;
};

export const convertDateFormat = (startDate: string): string => {
    return dateFormat(startDate, 'dd/mm/yyyy');
};

export const replaceInternalNocCode = (nocCode: string): string => {
    if (nocCode === INTERNAL_NOC) {
        return 'WBTR';
    }

    return nocCode;
};

let connectionPool: Pool;

const executeQuery = async <T>(query: string, values: string[]): Promise<T> => {
    if (!connectionPool) {
        connectionPool = getAuroraDBClient();
    }
    const [rows] = await connectionPool.execute(query, values);
    return JSON.parse(JSON.stringify(rows));
};

export const getServicesByNocCode = async (nocCode: string): Promise<ServiceType[]> => {
    const nocCodeParameter = replaceInternalNocCode(nocCode);
    logger.info('', {
        context: 'data.auroradb',
        message: 'retrieving services for given noc',
        noc: nocCode,
    });

    try {
        const queryInput = `
            SELECT lineName, startDate, serviceDescription AS description, serviceCode
            FROM tndsOperatorService
            WHERE nocCode = ?
            ORDER BY CAST(lineName AS UNSIGNED) = 0, CAST(lineName AS UNSIGNED), LEFT(lineName, 1), MID(lineName, 2), startDate;
        `;

        const queryResults = await executeQuery<ServiceType[]>(queryInput, [nocCodeParameter]);

        return (
            queryResults.map(item => ({
                lineName: item.lineName,
                startDate: convertDateFormat(item.startDate),
                description: item.description,
                serviceCode: item.serviceCode,
            })) || []
        );
    } catch (error) {
        throw new Error(`Could not retrieve services from AuroraDB: ${error.stack}`);
    }
};

export const getOperatorNameByNocCode = async (nocCode: string): Promise<OperatorNameType> => {
    const nocCodeParameter = replaceInternalNocCode(nocCode);
    logger.info('', {
        context: 'data.auroradb',
        message: 'retrieving operator name for given noc',
        noc: nocCode,
    });

    const queryInput = `
    SELECT operatorPublicName
    FROM nocTable
    WHERE nocCode = ?
    `;

    let queryResult: OperatorNameType[];

    try {
        queryResult = await executeQuery<OperatorNameType[]>(queryInput, [nocCodeParameter]);
    } catch (error) {
        throw new Error(`Could not retrieve operator name from AuroraDB: ${error.stack}`);
    }

    return queryResult[0];
};

export const batchGetOperatorNamesByNocCode = async (nocCodes: string[]): Promise<OperatorNameType[]> => {
    logger.info('', {
        context: 'data.auroradb',
        message: 'retrieving operator name for given noc',
    });

    try {
        const substitution = nocCodes.map(() => '?').join(',');
        const batchQuery = `
            SELECT operatorPublicName, nocCode
            FROM nocTable
            WHERE nocCode IN (${substitution})
        `;

        return executeQuery<OperatorNameType[]>(batchQuery, nocCodes);
    } catch (error) {
        throw new Error(
            `Error performing batch get for operator names against noc codes '${JSON.stringify(nocCodes)}': ${
                error.stack
            }`,
        );
    }
};

export const batchGetStopsByAtcoCode = async (atcoCodes: string[]): Promise<Stop[] | []> => {
    logger.info('', {
        context: 'data.auroradb',
        message: 'retrieving naptan info for atco codes',
    });

    try {
        const substitution = atcoCodes.map(() => '?').join(',');
        const batchQuery = `
            SELECT commonName, naptanCode, atcoCode, nptgLocalityCode, localityName, parentLocalityName, indicator, street
            FROM naptanStop
            WHERE atcoCode IN (${substitution})
        `;

        const queryResults = await executeQuery<NaptanInfo[]>(batchQuery, atcoCodes);

        return queryResults.map(item => ({
            stopName: item.commonName,
            naptanCode: item.naptanCode,
            atcoCode: item.atcoCode,
            localityCode: item.nptgLocalityCode,
            localityName: item.localityName,
            parentLocalityName: item.parentLocalityName ?? '',
            indicator: item.indicator,
            street: item.street,
        }));
    } catch (error) {
        throw new Error(
            `Error performing batch get for naptan info for stop list '${JSON.stringify(atcoCodes)}': ${error.stack}`,
        );
    }
};

export const getAtcoCodesByNaptanCodes = async (naptanCodes: string[]): Promise<NaptanAtcoCodes[]> => {
    logger.info('', {
        context: 'data.auroradb',
        message: 'retrieving atco codes by naptan codes',
    });

    const substitution = naptanCodes.map(() => '?').join(',');
    const atcoCodesByNaptanCodeQuery = `
        SELECT atcoCode, naptanCode FROM naptanStop
        WHERE naptanCode IN (${substitution})
    `;

    try {
        const queryResults = await executeQuery<NaptanAtcoCodes[]>(atcoCodesByNaptanCodeQuery, naptanCodes);

        return queryResults.map(item => ({ atcoCode: item.atcoCode, naptanCode: item.naptanCode }));
    } catch (error) {
        throw new Error(
            `Error performing queries for ATCO Codes using Naptan Codes '${JSON.stringify(naptanCodes)}': ${
                error.stack
            }`,
        );
    }
};

export const getServiceByNocCodeAndLineName = async (nocCode: string, lineName: string): Promise<RawService> => {
    const nocCodeParameter = replaceInternalNocCode(nocCode);
    logger.info('', {
        context: 'data.auroradb',
        message: 'retrieving service info for given noc and line name',
        noc: nocCode,
        lineName,
    });

    const serviceQuery = `
        SELECT os.operatorShortName, os.serviceDescription, os.lineName, pl.fromAtcoCode, pl.toAtcoCode, pl.journeyPatternId, pl.orderInSequence, nsStart.commonName AS fromCommonName, nsStop.commonName as toCommonName
        FROM tndsOperatorService AS os
        JOIN tndsJourneyPattern AS ps ON ps.operatorServiceId = os.id
        JOIN tndsJourneyPatternLink AS pl ON pl.journeyPatternId = ps.id
        LEFT JOIN naptanStop nsStart ON nsStart.atcoCode=pl.fromAtcoCode
        LEFT JOIN naptanStop nsStop ON nsStop.atcoCode=pl.toAtcoCode
        WHERE os.nocCode = ? AND os.lineName = ?
        ORDER BY pl.journeyPatternId ASC, pl.orderInSequence
    `;

    let queryResult: QueryData[];

    try {
        queryResult = await executeQuery(serviceQuery, [nocCodeParameter, lineName]);
    } catch (error) {
        throw new Error(`Could not get journey patterns from Aurora DB: ${error.stack}`);
    }

    const service = queryResult[0];

    // allows to get the unique journey's for the operator e.g. [1,2,3]
    const uniqueJourneyPatterns = queryResult
        .map(item => item.journeyPatternId)
        .filter((value, index, self) => self.indexOf(value) === index);

    const rawPatternService = uniqueJourneyPatterns.map(journey => {
        const filteredJourney = queryResult.filter(item => {
            return item.journeyPatternId === journey;
        });

        return {
            orderedStopPoints: [
                {
                    stopPointRef: filteredJourney[0].fromAtcoCode,
                    commonName: filteredJourney[0].fromCommonName,
                },
                ...filteredJourney.map((data: QueryData) => ({
                    stopPointRef: data.toAtcoCode,
                    commonName: data.toCommonName,
                })),
            ],
        };
    });

    if (!service || rawPatternService.length === 0) {
        throw new Error(`No journey patterns found for nocCode: ${nocCodeParameter}, lineName: ${lineName}`);
    }

    return {
        serviceDescription: service.serviceDescription,
        operatorShortName: service.operatorShortName,
        journeyPatterns: rawPatternService,
    };
};

export const getSalesOfferPackagesByNocCode = async (nocCode: string): Promise<SalesOfferPackage[]> => {
    logger.info('', {
        context: 'data.auroradb',
        message: 'retrieving sales offer packages for given noc',
        noc: nocCode,
    });

    try {
        const queryInput = `
            SELECT name, description, purchaseLocations, paymentMethods, ticketFormats
            FROM salesOfferPackage
            WHERE nocCode = ?
        `;

        const queryResults = await executeQuery<RawSalesOfferPackage[]>(queryInput, [nocCode]);

        return (
            queryResults.map(item => ({
                name: item.name,
                description: item.description,
                purchaseLocations: item.purchaseLocations.split(','),
                paymentMethods: item.paymentMethods.split(','),
                ticketFormats: item.ticketFormats.split(','),
            })) || []
        );
    } catch (error) {
        throw new Error(`Could not retrieve services from AuroraDB: ${error.stack}`);
    }
};

export const insertSalesOfferPackage = async (nocCode: string, salesOfferPackage: SalesOfferPackage): Promise<void> => {
    logger.info('', {
        context: 'data.auroradb',
        message: 'inserting sales offer package for given noc',
        noc: nocCode,
    });

    const purchaseLocations = salesOfferPackage.purchaseLocations.toString();
    const paymentMethods = salesOfferPackage.paymentMethods.toString();
    const ticketFormats = salesOfferPackage.ticketFormats.toString();

    const insertQuery = `INSERT INTO salesOfferPackage 
    (nocCode, name, description, purchaseLocations, paymentMethods, ticketFormats) 
    VALUES (?, ?, ?, ?, ?, ?)`;
    try {
        await executeQuery(insertQuery, [
            nocCode,
            salesOfferPackage.name,
            salesOfferPackage.description,
            purchaseLocations,
            paymentMethods,
            ticketFormats,
        ]);
    } catch (error) {
        throw new Error(`Could not insert sales offer package into the salesOfferPackage table. ${error.stack}`);
    }
};

export const getSearchOperators = async (searchText: string, nocCode: string): Promise<Operator[]> => {
    logger.info('', {
        context: 'data.auroradb',
        message: 'retrieving operators for given search text and noc',
        noc: nocCode,
        search: searchText,
    });

    const nocCodeParameter = replaceInternalNocCode(nocCode);

    const searchQuery = `SELECT nocCode, operatorPublicName FROM nocTable WHERE nocCode IN (
                             SELECT DISTINCT nocCode FROM tndsOperatorService WHERE regionCode IN (
                                SELECT DISTINCT regionCode FROM tndsOperatorService WHERE nocCode = ?)
                        ) AND operatorPublicName LIKE ?`;

    try {
        return await executeQuery<Operator[]>(searchQuery, [nocCodeParameter, `%${searchText}%`]);
    } catch (error) {
        throw new Error(`Could not retrieve operators from AuroraDB: ${error.stack}`);
    }
};
