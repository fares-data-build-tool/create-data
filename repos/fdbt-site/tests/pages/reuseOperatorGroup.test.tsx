import * as React from 'react';
import { shallow } from 'enzyme';
import {
    mockFieldSetForReuseOperatorGroup,
    mockFieldSetForReuseOperatorGroupWithErrorsIfRadioNotSelected,
    mockFieldSetForReuseOperatorGroupWithErrorsIfOptionNotSelected,
    getMockContext,
} from '../testData/mockData';
import { ErrorInfo } from '../../src/interfaces';
import ReuseOperatorGroup, { getFieldset, getServerSideProps } from '../../src/pages/reuseOperatorGroup';
import { getOperatorGroupsByNoc } from '../../src/data/auroradb';

jest.mock('../../src/data/auroradb');

describe('pages', () => {
    describe('reuseOperatorGroup', () => {
        it('should render correctly', () => {
            const tree = shallow(
                <ReuseOperatorGroup errors={[]} csrfToken="" fieldset={mockFieldSetForReuseOperatorGroup} />,
            );
            expect(tree).toMatchSnapshot();
        });

        it('should render errors when user does not select a radio button', () => {
            const tree = shallow(
                <ReuseOperatorGroup
                    errors={[{ errorMessage: 'Choose one of the options below', id: 'conditional-form-group' }]}
                    csrfToken=""
                    fieldset={mockFieldSetForReuseOperatorGroupWithErrorsIfRadioNotSelected}
                />,
            );
            expect(tree).toMatchSnapshot();
        });

        it('should render errors when user does not provide a group name', () => {
            const tree = shallow(
                <ReuseOperatorGroup
                    errors={[
                        {
                            errorMessage: 'Choose a premade operator group from the options below',
                            id: 'premadeOperatorGroup',
                        },
                    ]}
                    csrfToken=""
                    fieldset={mockFieldSetForReuseOperatorGroupWithErrorsIfOptionNotSelected}
                />,
            );
            expect(tree).toMatchSnapshot();
        });

        describe('getFieldset', () => {
            it('should return fieldsets with no errors when no errors are passed', () => {
                const fieldsets = getFieldset([], ['Best Ops']);
                expect(fieldsets).toEqual(mockFieldSetForReuseOperatorGroup);
            });

            it('should return fieldsets with radio errors when radio errors are passed', () => {
                const radioErrors: ErrorInfo[] = [
                    { errorMessage: 'Choose one of the options below', id: 'conditional-form-group' },
                ];
                const fieldsets = getFieldset(radioErrors, ['Best Ops']);
                expect(fieldsets).toEqual(mockFieldSetForReuseOperatorGroupWithErrorsIfRadioNotSelected);
            });

            it('should return fieldsets with input errors when input errors are passed', () => {
                const inputErrors: ErrorInfo[] = [
                    {
                        errorMessage: 'Choose a premade operator group from the options below',
                        id: 'premadeOperatorGroup',
                    },
                ];
                const fieldsets = getFieldset(inputErrors, ['Best Ops']);
                expect(fieldsets).toEqual(mockFieldSetForReuseOperatorGroupWithErrorsIfOptionNotSelected);
            });
        });

        describe('getServerSideProps', () => {
            it('redirects on to searchOperators if the user does not have any premade operator groups saved', async () => {
                (getOperatorGroupsByNoc as jest.Mock).mockImplementation(() => []);
                const writeHeadMock = jest.fn();
                const mockContext = getMockContext({
                    mockWriteHeadFn: writeHeadMock,
                });
                await getServerSideProps(mockContext);
                expect(writeHeadMock).toBeCalledWith(302, { Location: '/searchOperators' });
            });
        });
    });
});
