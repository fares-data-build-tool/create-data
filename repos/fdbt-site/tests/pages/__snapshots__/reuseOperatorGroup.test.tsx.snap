// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages reuseOperatorGroup should render coorectly when one operator group is specified 1`] = `
<TwoThirdsLayout
  description="Select Operator Group page of the Create Fares Data Service"
  errors={Array []}
  title="Select Operator Group - Create Fares Data Service"
>
  <CsrfForm
    action="/api/reuseOperatorGroup"
    csrfToken=""
    method="post"
  >
    <ErrorSummary
      errors={Array []}
    />
    <div
      className="govuk-form-group "
    >
      <fieldset
        aria-describedby="reuse-operator-group-page-heading"
        className="govuk-fieldset"
      >
        <legend
          className="govuk-fieldset__legend govuk-fieldset__legend--l"
        >
          <h1
            className="govuk-fieldset__heading"
            id="reuse-operator-group-page-heading"
          >
            Select an operator group
          </h1>
        </legend>
        <div
          className="govuk-warning-text"
        >
          <span
            aria-hidden="true"
            className="govuk-warning-text__icon govuk-!-margin-top-1"
          >
            !
          </span>
          <strong
            className="govuk-warning-text__text"
          >
            <span
              className="govuk-warning-text__assistive"
            >
              Warning
            </span>
            You can create new operator group in your
             
            <a
              className="govuk-link"
              href="/viewOperatorGroups"
            >
              operator settings.
            </a>
             
            <br />
            Don't worry you can navigate back to this page when you are finished.
          </strong>
        </div>
      </fieldset>
      <div
        className="card-row"
        id="individual-passengers"
      >
        <OperatorGroupCard
          defaultChecked={false}
          index={0}
          key="1"
          operatorGroup={
            Object {
              "id": 1,
              "name": "Regular Child",
              "operators": Array [
                Object {
                  "name": "First operator",
                  "nocCode": "FOP",
                },
              ],
            }
          }
        />
      </div>
    </div>
    <input
      className="govuk-button govuk-!-margin-right-2"
      id="continue-button"
      type="submit"
      value="Continue"
    />
    <a
      className="govuk-button govuk-button--secondary"
      href="/viewOperatorGroups"
    >
      Create new
    </a>
  </CsrfForm>
</TwoThirdsLayout>
`;

exports[`pages reuseOperatorGroup should render correctly 1`] = `
<TwoThirdsLayout
  description="Select Operator Group page of the Create Fares Data Service"
  errors={Array []}
  title="Select Operator Group - Create Fares Data Service"
>
  <CsrfForm
    action="/api/reuseOperatorGroup"
    csrfToken=""
    method="post"
  >
    <ErrorSummary
      errors={Array []}
    />
    <div
      className="govuk-form-group "
    >
      <fieldset
        aria-describedby="reuse-operator-group-page-heading"
        className="govuk-fieldset"
      >
        <legend
          className="govuk-fieldset__legend govuk-fieldset__legend--l"
        >
          <h1
            className="govuk-fieldset__heading"
            id="reuse-operator-group-page-heading"
          >
            Select an operator group
          </h1>
        </legend>
        <div
          className="govuk-warning-text"
        >
          <span
            aria-hidden="true"
            className="govuk-warning-text__icon govuk-!-margin-top-1"
          >
            !
          </span>
          <strong
            className="govuk-warning-text__text"
          >
            <span
              className="govuk-warning-text__assistive"
            >
              Warning
            </span>
            You can create new operator group in your
             
            <a
              className="govuk-link"
              href="/viewOperatorGroups"
            >
              operator settings.
            </a>
             
            <br />
            Don't worry you can navigate back to this page when you are finished.
          </strong>
        </div>
      </fieldset>
      <span
        className="govuk-body"
      >
        <i>
          You currently have no operator group
        </i>
      </span>
    </div>
    <a
      className="govuk-button govuk-button--secondary"
      href="/viewOperatorGroups"
    >
      Create new
    </a>
  </CsrfForm>
</TwoThirdsLayout>
`;

exports[`pages reuseOperatorGroup should render errors when user does not select a radio button 1`] = `
<TwoThirdsLayout
  description="Select Operator Group page of the Create Fares Data Service"
  errors={
    Array [
      Object {
        "errorMessage": "Choose an operator group from the options below",
        "id": "operatorGroup-0",
      },
    ]
  }
  title="Select Operator Group - Create Fares Data Service"
>
  <CsrfForm
    action="/api/reuseOperatorGroup"
    csrfToken=""
    method="post"
  >
    <ErrorSummary
      errors={
        Array [
          Object {
            "errorMessage": "Choose an operator group from the options below",
            "id": "operatorGroup-0",
          },
        ]
      }
    />
    <div
      className="govuk-form-group "
    >
      <fieldset
        aria-describedby="reuse-operator-group-page-heading"
        className="govuk-fieldset"
      >
        <legend
          className="govuk-fieldset__legend govuk-fieldset__legend--l"
        >
          <h1
            className="govuk-fieldset__heading"
            id="reuse-operator-group-page-heading"
          >
            Select an operator group
          </h1>
        </legend>
        <div
          className="govuk-warning-text"
        >
          <span
            aria-hidden="true"
            className="govuk-warning-text__icon govuk-!-margin-top-1"
          >
            !
          </span>
          <strong
            className="govuk-warning-text__text"
          >
            <span
              className="govuk-warning-text__assistive"
            >
              Warning
            </span>
            You can create new operator group in your
             
            <a
              className="govuk-link"
              href="/viewOperatorGroups"
            >
              operator settings.
            </a>
             
            <br />
            Don't worry you can navigate back to this page when you are finished.
          </strong>
        </div>
      </fieldset>
      <div
        className="card-row"
        id="individual-passengers"
      >
        <OperatorGroupCard
          defaultChecked={false}
          index={0}
          key="1"
          operatorGroup={
            Object {
              "id": 1,
              "name": "Regular Child",
              "operators": Array [
                Object {
                  "name": "First operator",
                  "nocCode": "FOP",
                },
              ],
            }
          }
        />
      </div>
    </div>
    <input
      className="govuk-button govuk-!-margin-right-2"
      id="continue-button"
      type="submit"
      value="Continue"
    />
    <a
      className="govuk-button govuk-button--secondary"
      href="/viewOperatorGroups"
    >
      Create new
    </a>
  </CsrfForm>
</TwoThirdsLayout>
`;
